File: frontend/types/trade.ts
/**
 * TypeScript type definitions for Trade entities.
 * Must match backend Pydantic schemas and Python enums (中文 values).
 */

// ── 枚举值（与后端 Python 枚举一致）──────────────────────────────

export const ACCOUNT_TYPES = ["短线账户", "中线账户"] as const;
export type AccountType = (typeof ACCOUNT_TYPES)[number];

export const TRADE_CYCLES = ["短线", "中线", "长线"] as const;
export type TradeCycle = (typeof TRADE_CYCLES)[number];

export const PNL_FLAGS = ["盈利", "亏损", "保本出局"] as const;
export type PnLFlag = (typeof PNL_FLAGS)[number];

export const MARKET_ENVIRONMENTS = [
  "牛市-主升", "牛市-调整", "熊市-主跌", "熊市-反弹",
  "震荡市-上轨", "震荡市-下轨", "震荡市-中枢",
] as const;
export type MarketEnvironment = (typeof MARKET_ENVIRONMENTS)[number];

export const SECTOR_STATUSES = ["启动期", "主升期", "高潮期", "退潮期", "混沌期"] as const;
export type SectorStatus = (typeof SECTOR_STATUSES)[number];

export const PLAN_ADHERENCES = ["完全按计划", "轻微偏离", "临盘起意冲动交易"] as const;
export type PlanAdherence = (typeof PLAN_ADHERENCES)[number];

export const STOP_LOSS_DISCIPLINES = [
  "有预设并严格执行", "有预设但未执行", "无预设止损",
] as const;
export type StopLossDiscipline = (typeof STOP_LOSS_DISCIPLINES)[number];

export const EXIT_TYPES = ["主动止盈", "被动止损", "保本出局", "模式外情绪化离场"] as const;
export type ExitType = (typeof EXIT_TYPES)[number];

export const PSYCHOLOGICAL_STATES = ["冷静", "追涨FOMO", "贪婪", "恐惧"] as const;
export type PsychologicalState = (typeof PSYCHOLOGICAL_STATES)[number];

export const RESULT_TYPES = ["正确盈利", "运气盈利", "执行亏损", "模式亏损"] as const;
export type ResultType = (typeof RESULT_TYPES)[number];

export const ERROR_LEVELS = ["执行层错误", "模式层错误", "环境层错误"] as const;
export type ErrorLevel = (typeof ERROR_LEVELS)[number];

export const SELECTION_DIMENSIONS = ["技术面", "政策面", "基本面", "事件驱动", "情绪接力"] as const;

export const STRATEGY_PATTERNS = [
  "突破买入", "回踩低吸", "龙头首阴", "打板追涨", "均线支撑",
] as const;

// ── Trade 实体 ─────────────────────────────────────────────────

export interface Trade {
  id: string;
  user_id: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  version: number;
  llm_analysis_status: string;
  llm_raw_log: Record<string, unknown> | null;

  account_type: string;
  stock_code: string;
  stock_name: string | null;
  trade_cycle: string;
  entry_date: string;
  exit_date: string | null;
  position_size: number;
  entry_price: number;
  exit_price: number | null;
  preset_stop_loss: number | null;
  preset_take_profit: number | null;
  slippage: number | null;
  max_favorable_excursion: number | null;
  max_adverse_excursion: number | null;
  pnl_amount: number | null;
  pnl_ratio: number | null;
  pnl_flag: string | null;

  market_environment: string | null;
  sector_status: string | null;
  selection_dimension: string[] | null;
  strategy_pattern: string[] | null;
  volume_profile: string | null;
  thesis_statement: string | null;

  plan_adherence: string | null;
  stop_loss_discipline: string | null;
  exit_type: string | null;
  exit_reason: string | null;
  psychological_state: string | null;

  result_type: string | null;
  error_level: string | null;
  environment_mismatch_flag: boolean | null;
  permanent_exclusion_flag: boolean | null;
  correct_action: string;
  llm_action_item: string | null;
}

// ── API 响应类型 ───────────────────────────────────────────────

export interface StandardResponse<T> {
  success: boolean;
  data: T;
  message: string;
  error?: { code: string; details?: unknown };
}

export interface PaginationInfo {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface PaginatedResponse<T> {
  success: boolean;
  data: T[];
  pagination: PaginationInfo;
  message: string;
}

export interface RulesSummary {
  total_trades: number;
  permanent_exclusions_count: number;
  correct_behaviors_count: number;
  environment_mismatches_count: number;
  execution_errors_count: number;
  pattern_errors_count: number;
  environment_errors_count: number;
}

// ── 创建/更新请求类型 ─────────────────────────────────────────

export interface TradeCreateData {
  account_type: AccountType;
  stock_code: string;
  stock_name?: string;
  trade_cycle: TradeCycle;
  entry_date: string;
  exit_date?: string;
  position_size: number;
  entry_price: number;
  exit_price?: number;
  preset_stop_loss?: number;
  preset_take_profit?: number;
  slippage?: number;
  max_favorable_excursion?: number;
  max_adverse_excursion?: number;
  market_environment?: MarketEnvironment;
  sector_status?: SectorStatus;
  selection_dimension?: string[];
  strategy_pattern?: string[];
  volume_profile?: string;
  thesis_statement?: string;
  plan_adherence?: PlanAdherence;
  stop_loss_discipline?: StopLossDiscipline;
  exit_type?: ExitType;
  exit_reason?: string;
  psychological_state?: PsychologicalState;
  result_type?: ResultType;
  error_level?: ErrorLevel;
  environment_mismatch_flag?: boolean;
  permanent_exclusion_flag?: boolean;
  correct_action: string;
}

export interface TradeUpdateData extends Partial<TradeCreateData> {
  version: number;
}

File: backend/app/schemas/trade.py
"""
Pydantic schemas for Trade API request/response validation.

所有枚举引用自 app/models/enums.py（database-schema 中定义的唯一数据源）。
"""

from datetime import datetime
from typing import Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import (
    AccountType,
    ErrorLevel,
    ExitType,
    MarketEnvironment,
    PlanAdherence,
    PnLFlag,
    PsychologicalState,
    ResultType,
    SectorStatus,
    StopLossDiscipline,
    TradeCycle,
)


# ── 创建请求 ────────────────────────────────────────────────────────


class TradeCreate(BaseModel):
    """创建交易记录请求 — 对应 PRD 中的四步表单提交"""

    # 模块 1: 基础交易事实
    account_type: AccountType
    stock_code: str = Field(min_length=1, max_length=20)
    stock_name: Optional[str] = Field(None, max_length=100)
    trade_cycle: TradeCycle
    entry_date: datetime
    exit_date: Optional[datetime] = None
    position_size: float = Field(gt=0, le=100, description="仓位占比 (%)")
    entry_price: float = Field(gt=0)
    exit_price: Optional[float] = Field(None, gt=0)
    preset_stop_loss: Optional[float] = None
    preset_take_profit: Optional[float] = None
    slippage: Optional[float] = Field(None, ge=0)
    max_favorable_excursion: Optional[float] = None
    max_adverse_excursion: Optional[float] = None

    # 模块 2: 决策环境快照
    market_environment: Optional[MarketEnvironment] = None
    sector_status: Optional[SectorStatus] = None
    selection_dimension: Optional[List[str]] = None
    strategy_pattern: Optional[List[str]] = None
    volume_profile: Optional[str] = Field(None, max_length=500)
    thesis_statement: Optional[str] = None

    # 模块 3: 执行与心理评估
    plan_adherence: Optional[PlanAdherence] = None
    stop_loss_discipline: Optional[StopLossDiscipline] = None
    exit_type: Optional[ExitType] = None
    exit_reason: Optional[str] = None
    psychological_state: Optional[PsychologicalState] = None

    # 模块 4: 深度归因
    result_type: Optional[ResultType] = None
    error_level: Optional[ErrorLevel] = None
    environment_mismatch_flag: Optional[bool] = False
    permanent_exclusion_flag: Optional[bool] = False
    correct_action: str = Field(min_length=1, description="正确行为（必填）")

    @model_validator(mode="after")
    def validate_dates(self) -> "TradeCreate":
        if self.exit_date and self.entry_date:
            if self.exit_date < self.entry_date:
                raise ValueError(
                    "exit_date（卖出时间）必须晚于 entry_date（买入时间）"
                )
        return self


# ── 更新请求 ────────────────────────────────────────────────────────


class TradeUpdate(BaseModel):
    """部分更新请求 — 所有字段可选，附带 version 用于乐观锁"""

    account_type: Optional[AccountType] = None
    stock_code: Optional[str] = Field(None, min_length=1, max_length=20)
    stock_name: Optional[str] = None
    trade_cycle: Optional[TradeCycle] = None
    entry_date: Optional[datetime] = None
    exit_date: Optional[datetime] = None
    position_size: Optional[float] = Field(None, gt=0, le=100)
    entry_price: Optional[float] = Field(None, gt=0)
    exit_price: Optional[float] = Field(None, gt=0)
    preset_stop_loss: Optional[float] = None
    preset_take_profit: Optional[float] = None
    slippage: Optional[float] = None
    max_favorable_excursion: Optional[float] = None
    max_adverse_excursion: Optional[float] = None
    market_environment: Optional[MarketEnvironment] = None
    sector_status: Optional[SectorStatus] = None
    selection_dimension: Optional[List[str]] = None
    strategy_pattern: Optional[List[str]] = None
    volume_profile: Optional[str] = None
    thesis_statement: Optional[str] = None
    plan_adherence: Optional[PlanAdherence] = None
    stop_loss_discipline: Optional[StopLossDiscipline] = None
    exit_type: Optional[ExitType] = None
    exit_reason: Optional[str] = None
    psychological_state: Optional[PsychologicalState] = None
    result_type: Optional[ResultType] = None
    error_level: Optional[ErrorLevel] = None
    environment_mismatch_flag: Optional[bool] = None
    permanent_exclusion_flag: Optional[bool] = None
    correct_action: Optional[str] = Field(None, min_length=1)

    # 乐观锁
    version: int = Field(description="当前版本号，用于并发控制")


# ── 响应模型 ────────────────────────────────────────────────────────


class TradeResponse(BaseModel):
    """交易记录响应模型 — 包含系统自动管理的字段"""

    # 模块 0: 系统字段
    id: UUID
    user_id: str
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None
    version: int
    llm_analysis_status: str
    llm_raw_log: Optional[dict] = None

    # 模块 1: 基础交易事实
    account_type: str
    stock_code: str
    stock_name: Optional[str] = None
    trade_cycle: str
    entry_date: datetime
    exit_date: Optional[datetime] = None
    position_size: float
    entry_price: float
    exit_price: Optional[float] = None
    preset_stop_loss: Optional[float] = None
    preset_take_profit: Optional[float] = None
    slippage: Optional[float] = None
    max_favorable_excursion: Optional[float] = None
    max_adverse_excursion: Optional[float] = None
    pnl_amount: Optional[float] = None
    pnl_ratio: Optional[float] = None
    pnl_flag: Optional[str] = None

    # 模块 2: 决策环境快照
    market_environment: Optional[str] = None
    sector_status: Optional[str] = None
    selection_dimension: Optional[List[str]] = None
    strategy_pattern: Optional[List[str]] = None
    volume_profile: Optional[str] = None
    thesis_statement: Optional[str] = None

    # 模块 3: 执行与心理评估
    plan_adherence: Optional[str] = None
    stop_loss_discipline: Optional[str] = None
    exit_type: Optional[str] = None
    exit_reason: Optional[str] = None
    psychological_state: Optional[str] = None

    # 模块 4: 深度归因
    result_type: Optional[str] = None
    error_level: Optional[str] = None
    environment_mismatch_flag: Optional[bool] = None
    permanent_exclusion_flag: Optional[bool] = None
    correct_action: str
    llm_action_item: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


# ── 查询参数 ────────────────────────────────────────────────────────


class TradeFilters(BaseModel):
    """查询过滤参数"""

    stock_code: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    market_environment: Optional[MarketEnvironment] = None
    result_type: Optional[ResultType] = None
    pnl_flag: Optional[PnLFlag] = None
    account_type: Optional[AccountType] = None
    trade_cycle: Optional[TradeCycle] = None


class PaginationParams(BaseModel):
    """分页参数"""

    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


class SortParams(BaseModel):
    """排序参数"""

    sort_by: str = Field(default="entry_date")
    order: str = Field(default="desc", pattern="^(asc|desc)$")


# ── 批量操作 ────────────────────────────────────────────────────────


class BulkCreateRequest(BaseModel):
    """批量创建请求"""

    trades: List[TradeCreate]


class BulkCreateResponse(BaseModel):
    """批量创建响应"""

    success_count: int
    failure_count: int
    successes: List[TradeResponse]
    failures: List[Dict]


# ── 规则库响应 ──────────────────────────────────────────────────────


class RulesSummary(BaseModel):
    """规则库汇总统计"""

    total_trades: int
    permanent_exclusions_count: int
    correct_behaviors_count: int
    environment_mismatches_count: int
    execution_errors_count: int
    pattern_errors_count: int
    environment_errors_count: int
